src/Span.ts

```diff
@@ -0,0 +1,375 @@
+import {
+  NullEventAction,
+  AuditEventAction,
+  LogEventAction,
+  TraceEventAction,
+  LogResponseStatus,
+  EventType,
+  TraceTags,
+  TypeEventMessage,
+  TypeEventMetadata,
+  TypeSpanContext,
+  EventTraceMetadata,
+  EventMessage,
+  TypeEventTypeAction,
+  EventStateMetadata,
+  EventMetadata,
+  TraceEventTypeAction,
+  AuditEventTypeAction,
+  LogEventTypeAction,
+  TypeEventAction
+} from './model/EventMessage'
+
+import {
+  IEventRecorder, DefaultLoggerRecorder, DefaultSidecarRecorder
+} from './Recorder'
+
+import { EventLoggingServiceClient } from './transport/EventLoggingServiceClient';
+
+const Config = require('./lib/config')
+
+type RecorderKeys = 'defaultRecorder' | 'logRecorder' | 'auditRecorder' | 'traceRecorder'
+
+const defaultRecorder = Config.SIDECAR_DISABLED
+  ? new DefaultLoggerRecorder()
+  : new DefaultSidecarRecorder(new EventLoggingServiceClient(Config.EVENT_LOGGER_SERVER_HOST, Config.EVENT_LOGGER_SERVER_PORT))
+
+
+type PartialWithDefaultRecorder<T> = {
+  [P in keyof T]?: T[P]
+} & {
+  defaultRecorder: IEventRecorder
+}
+
+type Recorders = PartialWithDefaultRecorder<{
+  defaultRecorder: IEventRecorder
+  logRecorder?: IEventRecorder,
+  auditRecorder?: IEventRecorder,
+  traceRecorder?: IEventRecorder
+}>
+
+
+// -===- SPAN -===-
+
+type TypeOfMessage = { [key: string]: NonNullable<any> } | string
+
+/**
+ * defines options to the injectContextToMessage and extractContextFromMessage
+ * @param type the carrier type
+ * @param path in the carrier where the trace context should be injected or extracted from
+ */
+type ContextOptions = {
+  type?: string,
+  path?: string
+}
+
+interface ISpan { // TODO abstract class
+  spanContext: TypeSpanContext
+  recorders: Recorders
+  info: (message: TypeOfMessage) => Promise<any>
```

Comment:
I think it would be a good idea to add JS docs to the interface please. I know you add it to the implementation below, but I think it would be good to also have the "generic" version on the interface, including any notes that someone may need when implementing it.